### ISSUE TITLE
Add option to marian decoder to generate alignment and attention matrices

### ISSUE THREAD

**Issue by emjotde** [marian-nmt/marian-dev#145]
Add option to marian decoder to generate alignment and attention matrices. Should follow the format in amun.

**Comment by emjotde** [comment]
Also Ondrej pointed out that is would be useful to have a attention matrix mode for the rescorer to do forced alignment. Any takers?

**Comment by snukky** [comment]
Hard alignments can be generated by `marian-decoder` or `marian-server `by using  the `--alignment` option.

**Comment by snukky** [comment]
`marian-scorer` also supports `--alignment` w/ and w/o `--n-best` option.

**Comment by snukky** [comment]
Marian produces hard alignments identical as Amun does, except we use Moses-style formatting, which has the source index first in each pair. My concern is if we should follow Amun heuristics for generating alignments. 

Attention matrices contain probabilities for all source and target tokens including EOS tokens. Amun skips the target EOS token when converts soft alignments to hard alignments, but allows the source EOS token being paired with any previous target word (example below). I don't know the motivation for this, but it seems odd. 

@emjotde Should we 1) output alignments for source and target EOS tokens, filter both (if so, when - 2a) before or 2b) after finding the most probable word), or 3) leave it as it is now? I would vote for option 1 or 2b.

_Example_
Input (13 tokens) and output (15 tokens):
```
input : re@@ connecting With the Very American Ideal That Labor Rights Are Human Rights
output: Wieder@@ ansch@@ aff@@ ung mit der Very American Ideal , die Arbeits@@ rechte 
    sind Menschenrechte
```
Amun (trg-src) and Marian (src-trg) alignments:
```
amun  : 0-0 1-1 2-13 3-13 4-2 5-4 6-4 7-5 8-6 9-7 10-7 11-8 12-9 13-10 14-11
marian: 0-0 1-1 2-4 4-5 4-6 5-7 6-8 7-9 7-10 8-11 9-12 10-13 11-14 13-2 13-3
```

In the example, do we need alignments with source token index 13?

**Comment by emjotde** [comment]
Instead of looking at amun, how is nematus handling this?

**Comment by snukky** [comment]
I added generation of attention weights (a.k.a soft alignments) in decoder and scorer.

The theano-based version of Nematus have an option to output attention matrices, and they do contain weights for source and target EOS tokens. I've changed the way how attention weights and word alignments are generated in Marian and now all include EOS tokens. I think this is the right way of producing alignments - in the least modified form.
